pkg/controller/core.oam.dev/v1alpha2/application/dispatch/dispatch.go

```diff
@@ -0,0 +1,262 @@
+/*
+Copyright 2021 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dispatch
+
+import (
+	"context"
+	"reflect"
+
+	"github.com/pkg/errors"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1beta1"
+	"github.com/oam-dev/kubevela/pkg/utils/apply"
+)
+
+// NewAppManifestsDispatcher creates an AppManifestsDispatcher.
+func NewAppManifestsDispatcher(c client.Client, appRev *v1beta1.ApplicationRevision) *AppManifestsDispatcher {
+	return &AppManifestsDispatcher{
+		c:          c,
+		applicator: apply.NewAPIApplicator(c),
+		appRev:     appRev,
+		gcHandler:  NewGCHandler(c, appRev.Namespace),
+	}
+}
+
+// AppManifestsDispatcher dispatch application manifests into K8s and record the dispatched manifests' references in a
+// resource tracker which is named by a particular rule: name = appRevision's Name + appRevision's namespace.
+// A bundle of manifests to be dispatched MUST come from the given application revision.
+type AppManifestsDispatcher struct {
+	c          client.Client
+	applicator apply.Applicator
+	gcHandler  GarbageCollector
+
+	appRev *v1beta1.ApplicationRevision
+	oldRT  *v1beta1.ResourceTracker
+	skipGC bool
+
+	appRevName string
+	namespace  string
+	newRTName  string
+	newRT      *v1beta1.ResourceTracker
+}
+
+// EnableGC return an AppManifestsDispatcher that always do GC after dispatching resources.
+// GC will calculate diff between the dispatched resouces and ones recorded in the given resource tracker.
+func (a *AppManifestsDispatcher) EnableGC(rt *v1beta1.ResourceTracker) *AppManifestsDispatcher {
```

Comment:
where is it used?

Reply:
- InplaceUpgrade in app controller.  
- Before ending of rollout, v1=>2, it may need to GC v1.